src/engine/GroupBy.h

```diff
@@ -244,12 +244,90 @@ class GroupBy : public Operation {
       size_t numAggregates, const IdTable& subresult, size_t columnIndex,
       LocalVocab* localVocab);
 
+  // Stores the map which associates Ids with vector offsets and
+  // the vectors containing the aggregation data.
+  struct HashMapAggregationData {
+    // Maps `Id` to vector offsets.
+    ad_utility::HashMapWithMemoryLimit<KeyType, ValueType> map_;
+    // Stores the actual aggregation data.
+    std::vector<std::vector<AverageAggregationData>> aggregationData_;
+
+    // If `id` is not in map, add an element to the `aggregationData_` vectors
+    // and store the index.
+    size_t getOrCreateIndex(Id id) {
+      auto [iterator, wasAdded] = map_.try_emplace(id);
+
+      if (wasAdded) {
+        iterator->second = aggregationData_.at(0).size();
+        for (auto& aggregation : aggregationData_) aggregation.emplace_back();
+      }
+
+      return iterator->second;
+    }
+
+    // Returns a vector containing the offsets for all ids of `ids`,
+    // inserting entries if necessary.
+    std::vector<size_t> getHashEntries(std::span<const Id> ids) {
```

Comment:
Move those definitions to the `.cpp` file.

Reply:
(all the function definitions from this class that work on more than one entry at least.)